src/Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -405,15 +416,33 @@ public void ValidateBlock(ContextInformation context)
                 // Calculate the consensus flags and check they are valid.
                 context.Flags = this.NodeDeployments.GetFlags(context.BlockValidationContext.ChainedBlock);
 
-                int lastCheckpointHeight = this.checkpoints.GetLastCheckpointHeight();
-                if (context.BlockValidationContext.ChainedBlock.Height > lastCheckpointHeight)
+                // Check whether to use checkpoint to skip block validation.
+                context.BlockValidationContext.SkipValidation = false; 
+                if (this.nodeSettings.UseCheckpoints)
+                {
+                    int lastCheckpointHeight = this.checkpoints.GetLastCheckpointHeight();
+                    context.BlockValidationContext.SkipValidation = context.BlockValidationContext.ChainedBlock.Height <= this.checkpoints.GetLastCheckpointHeight();
+                    if (context.BlockValidationContext.SkipValidation)
+                        this.logger.LogTrace("Block validation will be partially skipped due to block height {0} is not greater than last checkpointed block height {1}.", context.BlockValidationContext.ChainedBlock.Height, lastCheckpointHeight);
+                }
+
+                // Check whether to use assumevalid switch to skip validation.
+                if (!context.BlockValidationContext.SkipValidation && (this.nodeSettings.BlockAssumedValid != null))
+                {
+                    ChainedBlock assumeValidBlock = this.Chain.GetBlock(this.nodeSettings.BlockAssumedValid);
+                    context.BlockValidationContext.SkipValidation = (assumeValidBlock != null) && (context.BlockValidationContext.ChainedBlock.Height <= assumeValidBlock.Height);
```

Comment:
What I would prefer to see is this
```
context.SkipValidation = ...
```
Unless there are reasons to set 'SkipValidation' from the caller but I don't see that as being the case.
WDYT

Reply:
Yes I can do one big conditional. I thought it might be easier to understand like this but I'm not sure it is. It's still a bit ugly. :grin:

Reply:
That's fine actually. 
Another thing `(assumeValidBlock != null)` should this not throw if null?

Reply:
Yeah good question. So I guess the question is if a incorrect hash is specified do we want to throw an exception or ignore assume valid check? At a minimum I should at least log it.

Reply:
if assumevalid hash is not on the chain, all is fine! that's perfectly normal scenario and you don't even need to log it imho, because you would just add extra log for each block on the chain ...